Remove CBasicBridging + CASTBridging

hamishknight opened the pull request:
Merge CASTBridging and CBasicBridging with ASTBridging and BasicBridging respectively, allowing ASTGen to use C++ interop, and for both ASTGen and SwiftCompilerSources to share the same bridging code. There's more work that can be done here to improve the bridging (e.g we could potentially remove the `bridged` and `unbridged` functions in ASTBridging.cpp now), but I'm leaving that for a future PR as there's already enough here.

hamishknight commented:
@swift-ci please test

hyp commented on lib/ASTGen/CMakeLists.txt:
```diff
@@ -38,3 +38,31 @@ add_pure_swift_host_library(swiftASTGen STATIC
+
+set(compile_options
+  ${c_include_paths_args}
+  "SHELL: -Xfrontend -enable-experimental-cxx-interop"
```
is it possible to use the `-cxx-interoperability-mode=default` flag here instead?

hyp replied:
the existing experimental flag is deprecated

hamishknight replied:
I believe that flag was added with 5.9, so we would need to require a 5.9 host compiler for that (which would require updating CI first)

hyp replied:
I see, got it.

hamishknight replied:
Ended up making this conditional on 5.9, so we will use the new flag where we can

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented on include/swift/AST/ASTBridgingWrappers.def:
```diff
@@ -31,9 +31,8 @@
-// Each AST node subclass stores a non-null raw value.
 #ifndef DECL
-#define DECL(Id, Parent) AST_BRIDGING_WRAPPER_NONNULL(Id##Decl)
+#define DECL(Id, Parent) AST_BRIDGING_WRAPPER_NULLABLE(Id##Decl)
```
At this point I kind of wonder whether we're really gaining much from defining separate macros for can-be-nullable, and never-nullable, and should just stamp out nullable bridging wrappers for everything. My main worry with that initially was that it would unnecessarily pollute the type namespace, but at this point a good chunk of types need nullable wrappers anyways. In any case, I can explore that in a follow up.

hamishknight commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -380,7 +380,9 @@ extension ASTGenVisitor {
   func generate(_ node: InheritedTypeListSyntax?) -> BridgedArrayRef {
     guard let node else {
-      return .init()
+      // NOTE: You cannot do '.init()' here as that will produce garbage
+      // (rdar://116825963).
+      return .init(data: nil, numElements: 0)
```
Ah I should probably remove these footguns now 😄

hamishknight commented:
@swift-ci please test

bnbarham commented on lib/Sema/TypeCheckMacros.cpp:
```diff
@@ -1121,18 +1119,18 @@ evaluateFreestandingMacro(FreestandingMacroExpansion *expansion,
         expansion->getSourceRange().Start.getOpaquePointerValue(),
         &evaluatedSourceOut);
-    if (!evaluatedSourceOut.data)
+    if (!evaluatedSourceOut.get().data())
```
Not a huge fan of `get`. `unbridged` would match the bridging .cpp.

hamishknight replied:
Hmm yeah that seems fair enough

hamishknight replied:
Ah, I remember why I named it this way now, it's because it matches all the wrappers in SILBridging.h. If we want to change to `unbridged` we probably ought to change those too. Not sure if you have any strong opinions on this @eeckstein?

eeckstein replied:
no strong opinions. We can change it

hamishknight commented on include/swift/Basic/BasicBridging.h:
```diff
@@ -47,67 +62,257 @@ SWIFT_BEGIN_NULLABILITY_ANNOTATIONS
+
+struct BridgedArrayRef {
+  const void *_Nullable data;
+  size_t numElements;
```
Look again :) But yeah I should probably fix the others too

bnbarham replied:
Hah, nice. How far apart was your push and my hitting comment 😅? Doesn't have to be in this PR FWIW.

hamishknight replied:
Yeah I think I'll do this in a follow up

eeckstein reviewed (commented):
nice!

hamishknight commented on include/swift/Basic/BasicBridging.h:
```diff
@@ -47,67 +62,257 @@ SWIFT_BEGIN_NULLABILITY_ANNOTATIONS
 
+SWIFT_NAME("getter:BridgedStringRef.data(self:)")
+BRIDGED_INLINE 
+const uint8_t *_Nullable BridgedStringRef_data(BridgedStringRef str);
```
Because C++ interop doesn't currently support importing C++ methods as computed properties 😢

hamishknight replied:
rdar://117280662 if you're interested

eeckstein replied:
okay

hamishknight commented on include/swift/AST/ASTBridging.h:
```diff
@@ -82,14 +175,761 @@ class BridgedDiagnosticFixIt {
 SWIFT_END_NULLABILITY_ANNOTATIONS
 
+#ifndef PURE_BRIDGING_MODE
+// In _not_ PURE_BRIDGING_MODE, briding functions are inlined and therefore
```
Because the bridging functions can require C++ headers that we don't want to include in "pure bridging mode", since that can trigger the Windows and LLDB C++ interop issues. I can update the comment to explain this (FWIW I copied it from SILBridging.h)

eeckstein reviewed (approved):
lgtm!

hamishknight commented on lib/ASTGen/Sources/ASTGen/Bridge.swift:
```diff
@@ -69,17 +69,17 @@ extension BridgedSourceRange {
   }
 }
 
 /// Allocate a copy of the given string as a null-terminated UTF-8 string.
```
Yeah seems like it probably should

hamishknight commented:
@swift-ci please test

hamishknight commented:
Will address the rest of the feedback in a follow up

hamishknight committed:
[SwiftCompilerSources] NFC: Allow building Package.swift

Tweak the compiler flags such that it can be
built from within Xcode. This is only useful when
editing, but I find it very handy.

hamishknight committed:
[SwiftCompilerSources] NFC: Workaround a compiler bug

hamishknight committed:
[ASTGen] Enable C++ interop

hamishknight committed:
[ASTGen] Migrate onto some BasicBridging types

Migrate onto BridgedStringRef, and the
BasicBridging equivalents of BridgedSourceLoc
and BridgedArrayRef.

hamishknight committed:
[Basic] Improve some bridging APIs

Improve APIs for BridgedStringRef,
BridgedOwnedString, and BridgedSourceLoc.

hamishknight committed:
[ASTGen] Fix a couple of `@_cdecl` param types

These are `void *` on the C side.

hamishknight committed:
[AST] Move BridgedNominalTypeDecl + BridgedVarDecl to ASTBridging

And renaming OptionalBridgedVarDecl to
BridgedNullableVarDecl for consistency with the
existing nullable AST node wrappers.

hamishknight committed:
[AST] Merge BridgedDiagnosticEngine + BridgedDiagEngine

Introduce a macro that can stamp out wrapper
classes for underlying C++ pointers, and use
it to define BridgedDiagnosticEngine in
ASTBridging. Then, migrate users of
BridgedDiagEngine onto it.

hamishknight committed:
[AST] Remove BridgedVarDecl + BridgedNominalTypeDecl duplicates

These are now stamped out by the ASTBridgingWrappers.

hamishknight committed:
Remove CBasicBridging and CASTBridging

Merge with BasicBridging and ASTBridging
respectively. The changes here should be pretty
uncontroversial, I tried to keep it to just moving
code about.

hamishknight committed:
[AST] Abstract away a couple of bridging wrappers

Move these into ASTBridgingWrappers.def.

hamishknight committed:
[AST] Remove redundant typedefs

These aren't needed in C++

hamishknight committed:
[AST] Use consistent naming for bridging functions

Add a `Bridged` prefix to match the name being
exposed to Swift, and to match the other bridging
functions. Also while here, use `SWIFT_NAME` for
some bridging functions that were missing it.

hamishknight committed:
[Basic] Improve API of BridgedArrayRef + BridgedData

Remove the default constructor footgun present with
the struct implementations, and sprinkle some
`SWIFT_NAME` and bridging utilities to make them
nicer to work with.

hamishknight committed:
[AST] Cleanup ASTBridging + BasicBridging

Sort AST node bridging functions into Expr, Stmt,
Decl, and TypeRepr, and tweak the headers to use
`MARK`.

hamishknight committed:
Remove BridgingUtils.h headers

These are now empty, and it seems like generally
we ought to prefer putting the bridging logic on
the BridgedXXX wrappers.

hamishknight committed:
Address review feedback

hamishknight commented:
@swift-ci please test

hamishknight commented:
🚢
